src/main/ruby/truffleruby/core/array.rb

```diff
@@ -296,11 +296,45 @@ def dig(idx, *more)
     if Primitive.nil?(result) || more.empty?
       result
     else
-      raise TypeError, "#{result.class} does not have #dig method" unless result.respond_to?(:dig)
-      result.dig(*more)
+      Array.looping_dig(result, *more)
     end
   end
 
+  def self.looping_dig(obj, *idxs)
+    idxs.each_with_index do |idx, n|
+      return obj if Primitive.nil?(obj)
+      case obj
+      when Hash
+        if basic_method?(obj, :dig)
+          obj = obj[idx]
+          next
+        end
+      when Array
+        if basic_method?(obj, :dig)
+          obj = obj.at(idx)
+          next
+        end
+      when Struct
+        if basic_method?(obj, :dig)
+          obj = nil
+          begin
+            obj = self[idx]
+          rescue IndexError, NameError
+            nil # nothing found with key
+          end
+          next
+        end
+      end
```

Comment:
This feels a bit like a rather-hacky CRuby optimization where they hardcode and duplicate logic for common cases, but that then slows down other cases. Can't say I'm a fan of this approach in general.

Reply:
Should fall back to the same thing?

Reply:
This is the kind of inlining failure we see:

```
[engine] Inline start Object#subject1                                             |call diff     0.00|Recursion Depth      0|Explore/inline ratio     1.00|IR Nodes   1239|Frequency     1.00|Truffle Callees      1|Forced false|Depth      0
[engine] Inlined        Array#dig <split-725d1814>                                |call diff   -12.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes    272|Frequency     1.00|Truffle Callees      3|Forced false|Depth      1
[engine] Inlined          Array#empty?                                            |call diff    -2.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes     82|Frequency     1.00|Truffle Callees      1|Forced false|Depth      2
[engine] Inlined            Array#size                                            |call diff    -1.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes     24|Frequency     1.00|Truffle Callees      0|Forced false|Depth      3
[engine] Inlined          Kernel#respond_to? <split-21dd5a29>                     |call diff    -1.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes    146|Frequency     1.00|Truffle Callees      0|Forced false|Depth      2
[engine] Inlined          Array#dig <split-5b35b11>                               |call diff    -8.00|Recursion Depth      1|Explore/inline ratio      NaN|IR Nodes    272|Frequency     1.00|Truffle Callees      3|Forced false|Depth      2
[engine] Inlined            Array#empty?                                          |call diff    -2.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes     82|Frequency     1.00|Truffle Callees      1|Forced false|Depth      3
[engine] Inlined              Array#size                                          |call diff    -1.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes     24|Frequency     1.00|Truffle Callees      0|Forced false|Depth      4
[engine] Inlined            Kernel#respond_to? <split-3a3bc854>                   |call diff    -1.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes    146|Frequency     1.00|Truffle Callees      0|Forced false|Depth      3
[engine] Inlined            Array#dig <split-7806e632>                            |call diff    -4.00|Recursion Depth      2|Explore/inline ratio      NaN|IR Nodes    272|Frequency     1.00|Truffle Callees      3|Forced false|Depth      3
[engine] Inlined              Array#empty?                                        |call diff    -2.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes     82|Frequency     1.00|Truffle Callees      1|Forced false|Depth      4
[engine] Inlined                Array#size                                        |call diff    -1.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes     24|Frequency     1.00|Truffle Callees      0|Forced false|Depth      5
[engine] Inlined              Kernel#respond_to? <split-245c5577>                 |call diff    -1.00|Recursion Depth      0|Explore/inline ratio      NaN|IR Nodes    146|Frequency     1.00|Truffle Callees      0|Forced false|Depth      4
[engine] Cutoff               Array#dig                                           |call diff     0.00|Recursion Depth      3|Explore/inline ratio      NaN|IR Nodes      0|Frequency     1.00|Truffle Callees      4|Forced false|Depth      4
```

Reply:
My point was that if `dig` is called on a non-Array/Hash/Struct it will now go through quite a few extra checks.

Right that inlining might be not be ideal, would be good to mention in the PR description or issue.
The idea here is then to be able to compile arbitrary-deep dig for Array/Hash/Struct as a single compilation unit?

Reply:
Ok well we'll explore more and show some data.

We have some deep `dig` operations yes.